### ISSUE TITLE
Can't run arm-none-eabi-gdb through Ghidra in MacOS (M1)

### ISSUE THREAD

**Issue by jonreyno** [NationalSecurityAgency/ghidra#7217]
**Describe the bug**
When trying to launch "arm-none-eabi-gdb" via raw gdb or remote gdb, I'm presented with an empty terminal window. I can run this just fine outside of Ghidra in a regular terminal window and can connect to the target, but running through the "Configure and Launch" dropdown doesn't work.

Update: If I select "Keep" in the timeout error window and wait about 10m it DOES eventually run gdb, but the tracermi connection ends as soon as the timeout happens.

**To Reproduce**
1. Go to Debugger
2. Click on Configure and Launch dropdown the raw gdb or remote gdb
3. Enter "arm-none-eabi-gdb" in the gdb field (installed via brew and in my PATH so no absolute path needed)
4. Click launch. Observe the empty terminal window.

**Expected behavior**
Terminal window populates with information or errors at the very least

**Screenshots**
<img width="1710" alt="Screenshot 2024-11-23 at 2 45 25 PM" src="https://github.com/user-attachments/assets/65e503ca-4b35-400c-9dde-ed51dffa699b">

<img width="417" alt="Screenshot 2024-11-23 at 2 46 06 PM" src="https://github.com/user-attachments/assets/f05bdc42-c70e-4893-be9d-702bccd5ad65">

<img width="466" alt="Screenshot 2024-11-23 at 2 46 33 PM" src="https://github.com/user-attachments/assets/e4f9d815-1fe8-4321-ac35-9c29f11c7ad9">

**Environment (please complete the following information):**
 - OS: macOS 14.6.1 (23G93) (M1 Max)
 - Java Version: openjdk 21.0.5 2024-10-15 LTS
 - Ghidra Version: 11.2.1
 - Ghidra Origin: installed via brew

**Comment by d-millar** [comment]
If you navigate to the arm-none-eabi-gdb executable on your Mac (via the ... button on the right) or enter the full path in "gdb command", does it still fail?

**Comment by jonreyno** [comment]
Yes, just tried that and it’s the same result.

**Comment by d-millar** [comment]
Any messages in the "Debug Console" window or in the launching terminal if you run ghidra from ./support/ghidraDebug (vs ghidraRun)?

**Comment by d-millar** [comment]
Also, any messages at the top of the Terminal?  I ask because, in your screenshot, the Terminal appear to have scrolled down.

**Comment by jonreyno** [comment]
Nothing in the terminal scrolling up and in debug I see these in the terminal window I launched ghidradebug from:
```
DEBUG New Pty: /dev/ttys086 at (181,182)  (UnixPty.java:48) 
INFO  local Pty session. PID = 25176  (LocalProcessPtySession.java:34) 
```
Doesn't seem like errors though...

**Comment by jonreyno** [comment]
Tried copy/pasting the contents of that terminal window for the hell of it and it was just all spaces

**Comment by jonreyno** [comment]
Does Ghidra run the same terminal process as just running it directly or is it something different?

**Comment by d-millar** [comment]
Hmmm, I know on my M1 /usr/bin/bash doesn't exist.  You might try editing the raw-gdb.sh file in Ghidra/Debug/Debugger-agent-gdb/data/debugger-launchers if that's try for you too.  I usually change the first line to /bin/bash.  That said, that usually throws a pretty obvious error.   Terminal process should be the native one.

**Comment by jonreyno** [comment]
I don't think that's it because if I put a different command in place of gdb (or even just type skdfaksdf in there) it throws an error on that line of the .sh so it seems like it does run the script and just hangs when it launches that particular bin.

**Comment by jonreyno** [comment]
Hmmm... ok so I just let it run in the background a while and it seems like it eventually ran it! Weird... so I guess it hangs for a really long time and if you let it keep running eventually it runs.

**Comment by jonreyno** [comment]
Problem is that the tracermi connection has already ended so it doesn't help me much that it runs eventually... still would have to manually do everything afterwards

**Comment by d-millar** [comment]
If you say "Keep" in that pop-up window, the timeout should be ignored - not the case?

**Comment by jonreyno** [comment]
I can select "Keep" and the terminal keeps running (and eventually ran the bin), but the tracermi connection ends as soon as I select "Keep". Also not sure if I can repro it running again... trying to do it again and haven't had any luck yet. It's either a longer wait than I thought or something else is going on...

**Comment by jonreyno** [comment]
Ok, yea, the wait is just pretty long. Finally launched again. I do have these errors in the terminal window. Is this because the tracermi connection ended before it could run?
`Python Exception <class 'ConnectionRefusedError'>: [Errno 61] Connection refused
Error occurred in Python: [Errno 61] Connection refused
Not connected`

**Comment by jonreyno** [comment]
Oh actually I was wrong, the tracermi connection ends as soon as the timeout error popup shows up so even if I just sit at that popup and don't touch anything it still won't work...

**Comment by jonreyno** [comment]
Can I disable the timeout?

**Comment by d-millar** [comment]
The timeout doesn’t  control the connection, so the connection is dying for some other reason.  Am assuming the parameters you’re using work fine in arm-none-eabi-gdb run from the console.  Not sure what’s going on here….

**Comment by jonreyno** [comment]
Yes, they run just fine in a separate terminal window. Was able to attach the debugger and everything… just had to manually enter all the commands the script runs and start up the tracermi connection manually as well.

**Comment by d-millar** [comment]
Huh, so you successfully executed the “ghidra trace” commands as well - interesting.   A couple of options:  I can try to debug this if the target is something I can get a copy of.  Or you can download the Ghidra source, build it, and I can help you debug it.  Or we can wait until Monday and see if @nsadeveloper789  has any insights / things to try.  He’s frequently got better ideas than I do.

**Comment by d-millar** [comment]
Oh, I have one more idea - is is possible your target is running when you connect?  We've had some issues recently, particularly with embedded targets, where we connect but either the target continues to run on connect or the no event is sent on connect so Ghidra's idea of the state is incorrect.  Am assuming you've defaulted to using "starti" in the initial dialog, but maybe try executing "interrupt" or hitting Ctrl-C in the Terminal immediately after the Terminal launches.

**Comment by d-millar** [comment]
Sorry, ignore the "starti" comment - not relevant for raw- or remote-gdb, but I still think this might be the issue.

**Comment by jonreyno** [comment]
In this case my target is a remote device (a 3ds), but this issue happens before I even try to connect to it. Even running raw gdb without trying to connect to anything has the issue. I did try ctrl-c and entering “interrupt” in the terminal just in case, but nothing happened. I’ll look into building it then...

**Comment by jonreyno** [comment]
Ok, I’m all setup to debug with Eclipse now. Where do I need to add breakpoints to see what is going on?

**Comment by jonreyno** [comment]
I tried this, but I didn’t see any issues… stepped a few times and then it was done and the terminal was running. I discovered another clue though. It seems that arm-none-eabi-gdb IS in fact running immediately, but the process is getting hung up when it launches from the ghidra terminal. I see it in Activity Monitor taking up 100% of one of the cpus. It stays there for around 5-6m and then disappears and that is when the terminal window finally updates. Running from a standard terminal window it just instantly launches… maybe it’s trying to communicate with the terminal that launched it and failing? I did see that what ghidra launches is called a “pseudo terminal” so maybe there are differences there?

**Comment by d-millar** [comment]
Hmmm, the subtleties of the pseudo-terminal are, I'm afraid, a bit of a mystery to me.  @nsadeveloper789 will have more insight on this as he implemented them.  If at any point you find a generically available target exhibiting the same behavior, I can definitely take a stab at debugging this.  Maybe, let's see first what @nsadeveloper789 says.

**Comment by jonreyno** [comment]
Here's another interesting thing... the pid of arm-none-eabi-gdb is 1 higher than the terminal process is that expected?
<img width="962" alt="Screenshot 2024-11-24 at 4 58 20 PM" src="https://github.com/user-attachments/assets/192f80fe-a7e2-42c7-be4e-bb8de6ca97f1">

**Comment by d-millar** [comment]
I think that's expected, i.e. arm-none-eabi-gdb is launched from the shell, so I would expect arm-none-eabi-gdb's pid to be at least the terminal's+1.

**Comment by d-millar** [comment]
Does this target have a GUI component by any chance?  i.e. is it possible the remote display variables not getting passed through the terminal?

**Comment by jonreyno** [comment]
No, it’s just in the terminal.

**Comment by jonreyno** [comment]
I tried running it from the old debugger targets window and got this error. Maybe this is the reason and the new flow doesn't catch it?

<img width="962" alt="Screenshot 2024-11-24 at 4 58 20 PM" src="https://github.com/user-attachments/assets/eb0a8708-ea00-4c31-bd5d-999f19ee81a6">

**Comment by d-millar** [comment]
Yeah, I don't think that's related. There were  a lot of problems with the terminal in the old mode of running, which was one of the motivations for switching to the python-based variant.  I am running local-gdb.sh with arm-none-eabi-gdb and no target without a problem on my mac.  I think if the mi2 issue were still present for the python version that wouldn't work.

**Comment by d-millar** [comment]
OK, just another data point: I am able to connect and debug an armv7 game emulator using remote-gdb.sh (with the /bin/bash edit and the full path to arm-none-eabi-arm), so am definitely thinking (as, I think, you are) that the failures have to do somehow with the specifics of your target.

**Comment by jonreyno** [comment]
I don't think it has anything to do with my target because it also happens with raw gdb which isn't connecting to a target. Seems more likely that it's something with my machine that's configured differently from yours. Are you running the latest version of arm-none-eabi-gdb from brew? Maybe a different version of java?

**Comment by nsadeveloper789** [comment]
I suppose I should also explain about timeout and the Keep button. This is already in our internal ticketing system: When the launcher runs, Ghidra simultaneously listens on a socket, expecting a connection from the script. That connection's timeout is set according to the `#@timeout` header, or 10 seconds if not specified. When that timeout expires, we close the socket and display the error dialog. We recognize that closing the socket there is a little pre-emptive. The Keep button allows you to read the Terminal output and maybe diagnose things, but it doesn't (yet) afford the launcher any extra time.

**Comment by d-millar** [comment]
Alright, let's see how close I can get to your setup:

Sadly< I have no way to get the first thing to match, but I feel like it's unlikely to be the culprit:

OS:  macOS 15.1.1 M1 Max
Java Version: openjdk 21.0.3 2024-4-16 LTS (OK, also different but again unlikely)
Ghidra Version: 11.2.1
Ghidra Origin: installed from source
   built as described on the github.com/NationalSecurityAgency/ghidra, up to and including prepDev step
Ghidra modification:  Ghidra/Debug/Debugger-agent-gdb/data/debugger-launchers/local-gdb.sh 
   changed to use "#!/bin/bash"
python version: 3.9.20
bash version: GNU bash, version 3.2.57(1)-release (arm64-apple-darwin)
arm-none-eabi-gdb:  15.2, installed via brew

Experiment:
./Ghidra/RuntimeScripts/Linux/ghidraRun
fresh project
Configure and Launch... gdb

Image:
Arguments:
gdb command: /opt/homebrew/bin/arm-none-eabi-gdb
Run command: starti
Architecture: 
Inferior tty: unchecked

Terminal output:
```
GNU gdb (GDB) 15.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=aarch64-apple-darwin24.0.0 --target=arm-none-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:67: SyntaxWarning: invalid escape 
sequence '\['
  AVAILABLE_PATTERN = re.compile('Available\[(?P<pid>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:68: SyntaxWarning: invalid escape 
sequence '\['
  BREAKPOINT_PATTERN = re.compile('Breakpoints\[(?P<breaknum>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:69: SyntaxWarning: invalid escape 
sequence '\['
  BREAK_LOC_PATTERN = extre(BREAKPOINT_PATTERN, '\[(?P<locnum>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:70: SyntaxWarning: invalid escape 
sequence '\['
  INFERIOR_PATTERN = re.compile('Inferiors\[(?P<infnum>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:71: SyntaxWarning: invalid escape 
sequence '\.'
  INF_BREAKS_PATTERN = extre(INFERIOR_PATTERN, '\.Breakpoints')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:72: SyntaxWarning: invalid escape 
sequence '\.'
  ENV_PATTERN = extre(INFERIOR_PATTERN, '\.Environment')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:73: SyntaxWarning: invalid escape 
sequence '\.'
  THREADS_PATTERN = extre(INFERIOR_PATTERN, '\.Threads')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:74: SyntaxWarning: invalid escape 
sequence '\['
  THREAD_PATTERN = extre(THREADS_PATTERN, '\[(?P<tnum>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:75: SyntaxWarning: invalid escape 
sequence '\.'
  STACK_PATTERN = extre(THREAD_PATTERN, '\.Stack')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:76: SyntaxWarning: invalid escape 
sequence '\['
  FRAME_PATTERN = extre(STACK_PATTERN, '\[(?P<level>\\d*)\]')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:77: SyntaxWarning: invalid escape 
sequence '\.'
  REGS_PATTERN = extre(FRAME_PATTERN, '\.Registers')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:78: SyntaxWarning: invalid escape 
sequence '\.'
  MEMORY_PATTERN = extre(INFERIOR_PATTERN, '\.Memory')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:79: SyntaxWarning: invalid escape 
sequence '\.'
  MODULES_PATTERN = extre(INFERIOR_PATTERN, '\.Modules')
/Users/llero/git/ghidra/Ghidra/Debug/Debugger-agent-gdb/build/pypkg/src/ghidragdb/methods.py:80: SyntaxWarning: invalid escape 
sequence '\['
  MODULE_PATTERN = extre(MODULES_PATTERN, '\[(?P<modname>.*)\]')
Requires an argument. Valid arguments are arm, armv2, armv2a, armv3, armv3m, armv4, armv4t, armv5, armv5t, armv5te, xscale, iwm
mxt, iwmmxt2, armv5tej, armv6, armv6kz, armv6t2, armv6k, armv7, armv6-m, armv6s-m, armv7e-m, armv8-a, armv8-r, armv8-m.base, ar
mv8-m.main, armv8.1-m.main, armv9-a, arm_any, auto.
Connected to Ghidra 11.3 at 127.0.0.1:61504
ARM:LE:32:v8 not found in compiler map
(gdb) 
```

**Comment by d-millar** [comment]
So, I guess, where are we differing?

**Comment by jonreyno** [comment]
I can confirm that adding the timeout allowed everything to run as normal (aside from the enormous wait). I also see that I may have 2 installations of java (one though brew and one through an externally downloaded package) and they might be conflicting with each other so I'll try and resolve that later tonight and see if that is my issue. Will update the thread with my findings.

**Comment by d-millar** [comment]
Excellent - thank you!

**Comment by jonreyno** [comment]
How do you see the version of bash installed?

**Comment by d-millar** [comment]
"bash --version" for me 
which bash == /bin/bash

**Comment by jonreyno** [comment]
Ok mine is:
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin23)

... strange that mine has a "23" at the end but is otherwise identical.

**Comment by jonreyno** [comment]
Do you see this when you run bash from the terminal?

jonreyno@Jonathans-MBP ~ % bash

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.

**Comment by d-millar** [comment]
yes, same same

**Comment by jonreyno** [comment]
Ok, I cleaned up the other java installation, but it didn't make a difference. I did notice that my python is 3.12.2 though... not sure if that would make a difference?

**Comment by jonreyno** [comment]
We also differ here:

Yours:
This GDB was configured as "--host=aarch64-apple-darwin24.0.0 --target=arm-none-eabi".

Mine:
This GDB was configured as "--host=aarch64-apple-darwin**23.4.0** --target=arm-none-eabi".

**Comment by d-millar** [comment]
I think the 23/24 diffs are just a result of the fact that I'm running Sequoia.  Am not at my computer right now, but, when I am, I can certainly test against Python 3.12.  I generally use 12 or 13, but 9 is baked into the Sequoia release, so was defaulting to that one.

**Comment by jonreyno** [comment]
Just so we could rule OS version out, I upgraded to Sequoia. Unfortunately it didn't change anything...

**Comment by jonreyno** [comment]
Ok, I've just had a breakthrough! I went ahead and recompiled the latest Ghidra on Sequoia and lo and behold it ran the terminal immediately! So seems like the issue was a combination of OS + building latest. Now I need to solve this other issue where my debugger window is in a state where it no longer updates. None of the windows are populating and resizing doesn't refresh the windows either. I could only tell the terminal ran immediately due to the debug spew. By the way, I had gotten into this state even before I upgraded to Sequoia. Any ideas?

**Comment by jonreyno** [comment]
Is this going to limit what I can do with Ghidra btw? What will I not be able to do without a compiler map for the processor my target uses?
<img width="946" alt="Screenshot 2024-11-26 at 8 45 30 PM" src="https://github.com/user-attachments/assets/85e86989-0aa9-474a-9e5a-8e2a5aed6271">

**Comment by d-millar** [comment]
Hmmm, I just added that message - maybe that wasn't a good idea.  I should probably modify it so it's not a source of distress.  That said, it should have no negative impact at all AFAICT, i.e. walking the source, am pretty sure the compiler is just set to "default", which will end up treating your target as though it was compiled with gcc.  This mostly matters on Windows targets where the compiler actually affects the language model.

**Comment by jonreyno** [comment]
To follow up, it seems that gdb is only running for me on a debug build of Ghidra. I tried running via ghidrarun on the same build and it did the same hanging thing. Also... when I do run in debug and attach to my 3ds I think something is wrong because I don't get any trace info. Seems like it can't read the memory, but I can see the registers. Any ideas?

This is what it looks like when I first attach:
<img width="1481" alt="Screenshot 2024-11-28 at 10 45 21 PM" src="https://github.com/user-attachments/assets/05690e45-5605-4d9d-a8b1-b9932a5c3871">

This is what the terminal outputs if I resume then interrupt again:
<img width="945" alt="Screenshot 2024-11-28 at 10 48 36 PM" src="https://github.com/user-attachments/assets/d942474e-ee2c-4f91-8d50-77c3ce4853ac">

**Comment by d-millar** [comment]
If you go to the Regions view and from the pull-down at the top right choose "Force full-view", you should have memory visible in the Dynamic Listing.  To get the Dynamic Listing to track a program loaded in the Static Listing is a little more work.  I recommend perusing the discussion in #7176 or any of the other references to arm-none-eabi-gdb or gdb-multiarch in Issues. The basic problem is your gdbstub on the 3ds (probably) does not support the command "info proc mappings", so Ghidra doesn't know where anything lives or what memory is valid.  Luckily, it's very easy to fake out the command, and we supply a template for doing so.

The InvalidObjPathError messages are a little more disturbing.  Not sure I know what's going on there.  As in #7176, you might try using the command "set debug remote 1" at the first break to set RSP into verbose mode.  Might be more obvious what packet is triggering the error.

**Comment by jonreyno** [comment]
> * I've lost track of what configuration is at play in each of these. When you run via `./ghidraRun`, is this an official release (11.2.1), or did you do a full build with Gradle? What branch/revision?

It happens with the latest public release as well as a locally built version at latest. Strangely, debug was also not working on both the public release AND the locally built version when I ran on Sonoma, but after upgrading to Sequoia, debug on my locally built version works just fine (but ghidrarun still doesn't work).

> * In either case, since we have different behavior in debug vs ghidraRun, this is going to be difficult to track down. Next time it happens, run `jps` from a terminal to get Ghidra's PID, then `jstack` to dump that stacks of every thread. There will be a lot. My best guess is the Swing (AWT) Event thread and some other thread (probably the one performing the launch) will be in deadlock, and I'd like to know where.

I'm about to take a trip, but I will try this when I return in about a week.

> 
> re/2:
> 
> * Best option here is to debug from your IDE and set a breakpoint on `InvalidObjPathError`. I'd like to know what path it's trying to get. When it hits, it should be in `TraceRmiHandler#getObject(ObjPath path, boolean required)`. Inspect the `path` variable.
> * Also, take a look further up in the Terminal window and see if there are any other "Exception in batch operation" messages above those. It's possible a request to create the object failed earlier, which would cause these missing object (path) errors. Start at the top and find the first such error.

Will do, thanks! Not sure if it's related to these errors, but I did follow d-millar's advice and fake out the info proc mappings which made the memory visible, but I keep finding myself in a situation where Ghidra and gdb get out of sync. For example, Ghidra thinks we're paused, but gdb says we're still running. Once this happens my only option seems to be to disconnect and shutdown the 3ds since it still thinks it's being debugged.

> 
> re/3:
> 
> * @d-millar has set you in the right direction. Note that at the moment "Force Full View" is only a partial solution. The better way is to use the script template to fake them out.
> * The relevant part of that long conversation is around here: [gdb error: Assertion `current_thread_ != nullptr' failed #7176 (comment)](https://github.com/NationalSecurityAgency/ghidra/issues/7176#issuecomment-2476753697). You may need to read upward for some context.
> * You'll need `info proc mappings` to work to see dynamic memory.
> * To synchronize it with the static listing, may also want `maint info sections -all-objects` working. Depending on circumstances (e.g., use of ASLR or not) there may be easier methods to synchronize.

I did the info proc mappings stuff and that seems to be working. I'll look into the static listing stuff next week when I'm back. Thanks for the help!

**Comment by nsadeveloper789** [comment]
Thanks for the info. Definitely interested in figuring out the state-out-of-sync thing, perhaps when you return. It's one thing when gdb and the target seem out of sync, but if gdb and Ghidra disagree, then our integrations need work. Enjoy your trip!
